### PR TITLE
use poll() and rbd_poll_io_events to speed up io retrieval.

### PR DESCRIPTION
**Pull request by liupan1111**
Hi,
In Aug. 2015, a new function rbd_poll_io_events is added in Ceph. I use it to speed up io retrieval. I tested by using P3700 SSD and ceph_jewel(bluestore), find the iops and latency has 10% improvement.

### CODE REVIEW

File: engines/rbd.c [hunk 1]
```diff
@@ -29,6 +33,7 @@ struct rbd_data {
 	rbd_image_t image;
 	struct io_u **aio_events;
 	struct io_u **sort_events;
+    int fd; /* add for poll */
```
Comment (axboe) [inline engines/rbd.c]: You used spaces, please use tabs.

File: engines/rbd.c [hunk 2]
```diff
@@ -127,6 +135,7 @@ static int _fio_setup_rbd_data(struct thread_data *td,
 
 }
 
+#define EVENT_SOCKET_TYPE_EVENTFD 2 /* add for rbd poll */
```
Comment (axboe) [inline engines/rbd.c]: Why do we have to define this? If the library is new enough to have support for image notification, doesn't it also include this define in the header?

File: engines/rbd.c [hunk 3]
```diff
@@ -188,8 +197,26 @@ static int _fio_rbd_connect(struct thread_data *td)
 		log_err("rbd_open failed.\n");
 		goto failed_open;
 	}
+
+	/* add for rbd poll */
+    rbd->fd = eventfd(0, EFD_NONBLOCK);
+    if (rbd->fd < 0)
```
Comment (axboe) [inline engines/rbd.c]: Spaces again, and you are using mixed styles of braces. Please the starting { on the same line as the if statement.

File: engines/rbd.c [hunk 4]
```diff
@@ -188,8 +197,26 @@ static int _fio_rbd_connect(struct thread_data *td)
...
+	r = rbd_set_image_notification(rbd->image, rbd->fd, EVENT_SOCKET_TYPE_EVENTFD);
+	if (r < 0) {
+		log_err("rbd_set_image_notification failed.\n");
+		goto failed_notify;
+	}
+	
 	return 0;
 
+failed_notify:
+    close(rbd->fd);
```
Comment (axboe) [inline engines/rbd.c]: Spaces...

File: engines/rbd.c [hunk 5]
```diff
@@ -205,6 +232,13 @@ static void _fio_rbd_disconnect(struct rbd_data *rbd)
 	if (!rbd)
 		return;
 
+    /* close eventfd */
+    if (rbd->fd >= 0)
```
Comment (axboe) [inline engines/rbd.c]: Spaces and braces.

File: engines/rbd.c (hunk 2 above)
Comment (axboe) [inline engines/rbd.c]: enum {                                                                          
  EVENT_TYPE_PIPE = 1,                                                          
  EVENT_TYPE_EVENTFD = 2                                                        
}; 

So it is there. I'd propose adding a configure check for whether set_image_type(..., EVENT_TYPE_EVENTFD) works or not, and use it in rbd.c if that is the case.

File: engines/rbd.c (hunk 1 above)
Comment (liupan1111) [inline engines/rbd.c]: Thanks. I will take care coding style next time.

One suggestion, how about adding a shell to check the coding style in fio package? If you agree, I could give help, draft a "check_format.sh", and let you review.

File: engines/rbd.c (hunk 2 above)
Comment (liupan1111) [inline engines/rbd.c]: I agree, please take a look at my new commit.

**Comment by axboe** [comment]
This looks better. Now please collapse the three patches into one proper patch, I don't want to pull your intermediate breakage into my history.

**Comment by liupan1111** [comment]
ok, I will close this pull request and create a new one with final changes.

**Comment by axboe** [comment]
Great, thanks!

**Pull request closed without merging**